src/ring_theory/unique_factorization_domain.lean

```diff
@@ -292,49 +337,198 @@ def of_unique_irreducible_factorization {α : Type*} [integral_domain α]
+            (dvd_iff_dvd_of_rel_right hfa.2.symm).1
               (multiset.dvd_prod hqa))
           (λ hqb, or.inr $ (dvd_iff_dvd_of_rel_left hq).2 $
-            (dvd_iff_dvd_of_rel_right (o.factors_prod hb0)).1
-              (multiset.dvd_prod hqb))⟩,
-  ..o }
+            (dvd_iff_dvd_of_rel_right hfb.2.symm).1
+              (multiset.dvd_prod hqb))
+        end⟩, irreducible_of_prime⟩
+
+theorem unique_factorization_monoid.of_exists_unique_irreducible_factors
+  [comm_cancel_monoid_with_zero α]
+  (eif : ∀ (a : α), a ≠ 0 → ∃ f : multiset α, (∀b ∈ f, irreducible b) ∧ a ~ᵤ f.prod)
+  (uif : ∀ (f g : multiset α),
+  (∀x∈f, irreducible x) → (∀x∈g, irreducible x) → f.prod ~ᵤ g.prod → multiset.rel associated f g) :
+  unique_factorization_monoid α :=
+unique_factorization_monoid_of_exists_prime_factors (by {
+  convert eif,
+  simp_rw irreducible_iff_prime_of_exists_unique_irreducible_factors eif uif,
+})
+
+namespace unique_factorization_monoid
+variables [comm_cancel_monoid_with_zero α] [decidable_eq α] [nontrivial α] [normalization_monoid α]
+  [unique_factorization_monoid α]
+
+/-- Noncomputably determines the multiset of prime factors. -/
+noncomputable def factors (a : α) : multiset α := if h : a = 0 then 0 else
+multiset.map normalize $ classical.some (unique_factorization_monoid.exists_prime_factors a h)
+
+theorem factors_prod {a : α} (ane0 : a ≠ 0) : associated (factors a).prod a :=
+begin
+  rw [factors, dif_neg ane0],
+  transitivity, swap,
+  { exact (classical.some_spec (unique_factorization_monoid.exists_prime_factors a ane0)).2.symm },
+  { rw [← associates.mk_eq_mk_iff_associated, ← associates.prod_mk, ← associates.prod_mk,
+    multiset.map_map],
+    refine congr rfl (congr (congr rfl _) rfl), ext,
+    rw [function.comp_apply, associates.mk_eq_mk_iff_associated], apply normalize_associated },
+end
+
+theorem prime_factors {a : α} : ∀ (x : α), x ∈ factors a → prime x :=
+begin
+  rw [factors],
+  by_cases ane0 : a = 0, {simp [ane0]},
+  rw dif_neg ane0,
+  intros x hx, rcases multiset.mem_map.1 hx with ⟨y, ⟨hy, rfl⟩⟩,
+  rw prime_iff_of_associated (normalize_associated),
+  exact (classical.some_spec (unique_factorization_monoid.exists_prime_factors a ane0)).1 y hy,
+end
+
+theorem irreducible_factors {a : α} : ∀ (x : α), x ∈ factors a → irreducible x :=
+λ x h, irreducible_of_prime (prime_factors x h)
+
+theorem normalize_factors {a : α} : ∀ (x : α), x ∈ factors a → normalize x = x :=
+begin
+  rw [factors],
+  by_cases a = 0, {simp [h]}, rw dif_neg h,
+  intros x hx, rcases multiset.mem_map.1 hx with ⟨y, ⟨hy, rfl⟩⟩,
+  apply normalize_idem,
+end
+
+lemma factors_irreducible {a : α} (ha : irreducible a) :
+  factors a = normalize a :: 0 :=
+begin
+  cases prime_factors_irreducible ha ⟨prime_factors, (factors_prod ha.ne_zero).symm⟩ with p hp,
+  convert hp.2, rw [← normalize_factors p, normalize_eq_normalize_iff, dvd_dvd_iff_associated],
+  {apply hp.1}, rw hp.2, simp,
+end
```

Comment:
The structure of a proof is hard to read if you put a lot of tactic commands on one line. I moved everything to separate lines and golfed a bit:

```suggestion
begin
  obtain ⟨p, a_assoc, hp⟩ := prime_factors_irreducible ha
    ⟨prime_factors, (factors_prod ha.ne_zero).symm⟩,
  have p_mem : p ∈ factors a,
  { rw hp, apply multiset.mem_cons_self },
  convert hp,
  rwa [← normalize_factors p p_mem, normalize_eq_normalize_iff, dvd_dvd_iff_associated]
end
```